src/sql/src/plan/query.rs

```diff
@@ -147,39 +147,44 @@ pub fn plan_insert_query(
         bail!("cannot insert into system table '{}'", table.name());
     }
 
+    // Normalize column names
+    let columns: Vec<_> = columns.into_iter().map(normalize::column_name).collect();
+
     // Validate target column order.
-    let mut project_key: Vec<_> = (0..desc.arity()).collect();
-    let mut ordering = Vec::with_capacity(desc.arity());
+    let mut source_types = Vec::with_capacity(columns.len());
+    let mut ordering = Vec::with_capacity(columns.len());
 
     if columns.is_empty() {
+        // Columns in source query must be in order. Let's guess the full shape and truncate to the
+        // right size later after planning the source query
+        source_types.extend(desc.iter_types().map(|x| &x.scalar_type));
         ordering.extend(0..desc.arity());
     } else {
-        let column_to_index: HashMap<&ColumnName, usize> = desc
-            .iter_names()
-            .filter_map(|x| x)
+        let column_by_name: HashMap<&ColumnName, (usize, &ColumnType)> = desc
+            .iter()
+            .filter_map(|(name, typ)| name.map(|n| (n, typ)))
             .enumerate()
-            .map(|(idx, name)| (name, idx))
+            .map(|(idx, (name, typ))| (name, (idx, typ)))
             .collect();
 
-        for (src_idx, c) in columns.into_iter().enumerate() {
-            let c = normalize::column_name(c);
-            if let Some(idx) = column_to_index.get(&c) {
+        let mut seen = HashSet::with_capacity(columns.len());
+
+        for c in &columns {
+            if let Some((idx, typ)) = column_by_name.get(c) {
                 ordering.push(*idx);
-                project_key[*idx] = src_idx;
+                source_types.push(&typ.scalar_type);
             } else {
                 bail!(
-                    "INSERT statement specifies column {}, but it is not present in table",
+                    "column \"{}\" of relation \"{}\" does not exist",
```

Comment:
Nice change! I like including the name of the relation here.

Reply:
Thanks!